kart/parse_args.py

```diff
@@ -0,0 +1,163 @@
+            #    $ git branch -c -- -x
+            #    fatal: '-x' is not a valid branch name.
+            # So we can assume it's a CLI flag, presumably for git rather than kart.
+            # It *could* be a path, but in that case the user should add a `--` before this option
+            # to disambiguate, and they haven't done so here.
+            issue_link = "https://github.com/koordinates/kart/issues/508"
+            warnings.warn(
+                f"{arg!r} is unknown to Kart and will be passed directly to git. "
+                f"This will be removed in Kart 0.12! Please comment on {issue_link} if you need to use this option.",
+                RemovalInKart012Warning,
+            )
+            return ArgType.OPTION
+        else:
+            raise click.UsageError(f"No such option: {arg}")
+
+    if allow_commits:
+        if arg == "[EMPTY]" or ".." in arg:
+            return ArgType.COMMIT
+
+        try:
+            repo.resolve_refish(arg)
+            return ArgType.COMMIT
+        except (KeyError, ValueError, pygit2.InvalidSpecError):
+            pass
+
+    if allow_filters:
+        return ArgType.FILTER
+
+    raise click.UsageError(
+        f"Argument not recognised as a valid commit, ref, or range: {arg}"
+    )
+
+
+def parse_commits_and_filters(
+    repo,
+    args,
+    kwargs=None,
+    allow_options=False,
+):
+    """
+    Interprets positional args for kart diff, show, and log, including "--", commits/refs/ranges, and filters.
+    Returns a three-tuple: (options, commits/refs/ranges, filters)
+    """
+
+    # As soon as we encounter a filter, we assume all remaining args are also filters.
+    # i.e. the filters must be given *last*.
+    # If it's ambiguous whether something is a filter or not, we assume it's a commit-ish.
+    # If you want to be unambiguous, provide the `--` arg to separate the list of commit-ish-es and filters.
+    # This behaviour should be consistent with git's behaviour.
+
+    if "--" in args:
+        dash_index = args.index("--")
+        filters = list(args[dash_index + 1 :])
+        args = args[:dash_index]
+    else:
+        dash_index = None
+        filters = []
+
+    options = []
+    commits = []
+
+    lists_by_type = {
+        ArgType.OPTION: options,
+        ArgType.COMMIT: commits,
+        ArgType.FILTER: filters,
+    }
+
+    allow_commits = True
+    allow_filters = dash_index is None
+    for arg in args:
+        arg_type = get_arg_type(
+            repo,
+            arg,
+            allow_options=allow_options,
+            allow_commits=allow_commits,
+            allow_filters=allow_filters,
+        )
+        lists_by_type[arg_type].append(arg)
+        if arg_type == ArgType.FILTER:
+            allow_commits = False
+
+    if kwargs is not None and allow_options:
+        for option_name, option_val in kwargs.items():
+            option_name = option_name.replace("_", "-", 1)
+            mapping = {
```

Comment:
this thing seems like it should just be a if/elif/else, and ditch the lambdas

Reply:
Done